lib/task.js

```diff
@@ -189,47 +198,108 @@ Task.prototype = {
     }
 
     return header + '\r\n';
-  },
+  }
 
-  logFooter: function(success, exitCode, start, finish) {
+  logFooter(success, exitCode, start, finish) {
     // Human readable success/failure thing...
-    var humanSuccess = success ?
+    let humanSuccess = success ?
       'Successful' :
       'Unsuccessful';
 
     // Yes, date subtraction yields a Number.
-    var duration = (finish - start) / 1000;
+    let duration = (finish - start) / 1000;
 
     return this.fmtLog(
       '%s task run with exit code: %d completed in %d seconds',
       humanSuccess, exitCode, duration
     );
-  },
+  }
 
-  logSchemaErrors: function(prefix, errors) {
+  logSchemaErrors(prefix, errors) {
     return this.fmtLog(
       "%s format is invalid json schema errors:\n %s",
       prefix, JSON.stringify(errors, null, 2)
     );
-  },
-
-  completeRun: function* (success) {
-    yield this.runtime.stats.timeGen(
-      'tasks.time.completed',
-      this.runtime.queue.reportCompleted(
-        this.status.taskId, this.runId, { success: success }
-      )
-    );
-  },
+  }
+
+  /**
+   * Aborts a run that is currently running or being prepared to run (pulling images,
+   * establishing states, etc).  This will optionally write an error to the stream
+   * and then write the footer and kill states.
+   *
+   * @param {String} stat - Name of the current state to be used when generating stats for killing states
+   * @param {String} error - Option error to write to the stream prior to aborting
+   */
+  async abortRun(stat, error) {
+    if (error) {
+      this.stream.write(error);
+    }
+    //
+    // Ensure that the stream has completely finished.
+    await this.stream.end(this.logFooter(
+      false, // unsuccessful task
+      -1, // negative exit code indicates infrastructure errors usually.
+      this.taskStart, // duration details...
+      new Date()
+    ));
+
+    await this.runtime.stats.timeGen(`tasks.time.states.${stat}.killed`, this.states.killed(this));
+  }
+
+  /**
+   * Resolves a run and reports the completion to the proper exchange.
+   *
+   * If a task was aborted because of a malformed payload or worker shutdown,
+   * the task will be reported as an exception.
+   *
+   * No Action will be taken for a canceled task because it has already been canceled
+   * by an external entity. Canceling the task run within the worker allows resources
+   * to be cleaned up.
+   */
+  async completeRun(success) {
+    let queue = this.runtime.queue;
+
+    if (this.taskException) {
+      if (this.taskState !== 'canceled') {
+        await this.runtime.stats.timeGen(
```

Comment:
Do we really want statistics for all calls to the queue?
If so, which might be valid... Can we not just make a hack and wrap the methods on the `taskcluster.Queue` instance?
So we don't have to read about it anymore...
Example:
```js
let queue = new taskcluster.Queue(...);
this.runtime.stats.wrapQueueClient(queue);
....
/** Wrap methods of a taskcluster client library */
function wrapQueueClient(client) {
  ['reportCompleted', 'reportException', ...].forEach((name) => {
    let method = client[name];
    client[name] = function(...args) {
      return timeGen('tasks.time.' + name, method.apply(this, args);
    };
  });
};
```

I'm also open to adding support for statistics collection to `taskcluster-client`. The client object to just emit an event whenever a request is started and whenever one is finished. The finished one could even include timestamp, if you would like. Anyways, abstracting this away with monkey patching is better than having it spread through the code base.